### PR TITLE
Reduxification: replace productsList with Redux counterpart

### PR DESCRIPTION
**Pull request by lamosty**
Part of solving #19828.

`lib/products-list` was reduxified. However, not all occurrences were replaced by its Redux counterpart. 

In this PR, we get rid of `lib/products-list`.

## Testing

Review commit by commit, testing all the affected files and flows.

### CODE REVIEW

File: client/lib/cart/store/index.js [hunk 1]
```diff
@@ -71,8 +72,24 @@ function emitChange() {
 }
 
 function update( changeFunction ) {
+	const reduxState = reduxGetState();
+	const isFetchingProductsList = isProductsListFetching( reduxState );
+	const productsList = getProductsList( reduxState );
+
+	if ( ! productsList && ! isFetchingProductsList ) {
```
Comment (samouri) [inline client/lib/cart/store/index.js]: this logic might be simpler as:
```js
if ( ! productsList ) {
  if ( ! isFetchingProductsList ) {
    reduxDispatch( requestProductsList );
  }
  setTimeout( () => update( changeFunction ), 300 );
}
```

File: client/lib/upgrades/actions/free-trials.js [hunk 2]
```diff
@@ -1,48 +0,0 @@
-/** @format */
```
Comment (samouri) [inline client/lib/upgrades/actions/free-trials.js]: why is this file deleted?

Comment (lamosty) [reply client/lib/upgrades/actions/free-trials.js]: We don't do free trials anymore. It's not used anywhere.

**Comment by lamosty** [comment]
I've rebased the PR. Could I get some more eyes, please? /cc @yurynix @scruffian @samouri @jsnajdr

**Comment by samouri** [comment]
@lamosty: I'm not very familiar with the affected area of code. Which flows should I test?
My guess is:
 - domain management
 - purchase flow
 - add gapps
 - signup

**Comment by klimeryk** [comment]
Seems like there's some duplication of efforts with @aidvu's https://github.com/Automattic/wp-calypso/pull/23329 ?

**Comment by kwight** [comment]
Is this still on the go @lamosty ? It could use a rebase, and then maybe an `e2e testing` label since it touches so much?

**Comment by lamosty** [comment]
> Is this still on the go @lamosty ? It could use a rebase, and then maybe an e2e testing label since it touches so much?

Sorry, I don't have resources to work on this anymore. We could either close or somebody could continue.

**Comment by jsnajdr** [comment]
I reassigned this PR to myself and rebased against latest master. There were some changes that already landed in @aidvu's #23329, so the new version of the PR is somewhat smaller.

File: client/lib/cart/store/index.js [hunk 3]
```diff
@@ -71,8 +72,22 @@ function emitChange() {
...
+	const reduxState = reduxGetState();
+	const isFetchingProductsList = isProductsListFetching( reduxState );
+	const productsList = getProductsList( reduxState );
+
+	if ( ! productsList ) {
+		if ( ! isFetchingProductsList ) {
+			reduxDispatch( requestProductsList );
+		}
+
+		setTimeout( () => update( changeFunction ), 300 );
```
Comment (blowery) [inline client/lib/cart/store/index.js]: why a timeout? could we hook into the reduxStore's change event instead?

File: client/components/data/domain-management/index.jsx [hunk 4]
```diff
@@ -35,66 +35,62 @@ function getStateFromStores( props ) {
...
-	},
+	}
 
-	render: function() {
+	render() {
 		return (
-			<div>
+			<Fragment>
 				<PageViewTracker path={ this.props.analyticsPath } title={ this.props.analyticsTitle } />
+				<QueryProducts />
```
Comment (blowery) [inline client/components/data/domain-management/index.jsx]: should we gate these on whether or not we have a valid products list? Or do we want to requery on every render when we're not already fetching?

File: client/signup/steps/domains/index.jsx [hunk 5]
```diff
@@ -353,17 +338,20 @@ class DomainsStep extends React.Component {
...
-				backUrl={ backUrl }
-				positionInFlow={ this.props.positionInFlow }
-				signupProgress={ this.props.signupProgress }
-				subHeaderText={ translate( "First up, let's find a domain." ) }
-				fallbackHeaderText={ translate( "Let's give your site an address." ) }
-				fallbackSubHeaderText={ fallbackSubHeaderText }
-				stepContent={ content }
-			/>
+			<Fragment>
+				<QueryProducts />,
```
Comment (blowery) [inline client/signup/steps/domains/index.jsx]: do we need the `,` here?

File: client/components/data/domain-management/index.jsx (hunk 4 above)
Comment (jsnajdr) [reply client/components/data/domain-management/index.jsx]: The Flux store only ever requests the data from network once per app lifetime. And persists them to local storage to make sure they are available immediately on app reboot.

I'll update `QueryProducts` to behave the same way. I believe the developer should only need to declare `<QueryProducts />` and let the implementation be smart enough to know when to issue a network request and when not.

File: client/signup/steps/domains/index.jsx (hunk 5 above)
Comment (jsnajdr) [reply client/signup/steps/domains/index.jsx]: Of course we don't, will fix 😀

File: client/lib/cart/store/index.js (hunk 3 above)
Comment (jsnajdr) [reply client/lib/cart/store/index.js]: The legacy `productsList` usage didn't wait for the list to update. It just triggered the request (by calling `get()`) and worked with the potentially empty object it returned.

@lamosty What was your intent here? To fix a bug?

**Comment by jsnajdr** [comment]
I pushed a few changes:
- removed the extra comma from JSX markup in `signup/steps/domains`. It was really visible when selecting a domain during signup starting from `/start/domains` 😨 

- patched `QueryProducts` so that the products list is requested only once per app lifetime.

- implemented the original `productsList.get()` logic in Cart Store's `update` function. If product list is empty, no extra product attributes will be added to cart item values. That potentially buggy behavior has been there for last 3 years and we don't want into the rabbit hole of fixing up the Cart Store. We'll get to that during the Cart Redux project.

- changed the `productsList` Redux persistence schema to be more permissive about product's `cost` field. If we don't this, persisted products list data will be rejected when trying to populate the initial Redux state with them.

Ready for another round of review now.

**Comment by jsnajdr** [comment]
Flows to test:
- domain management: when adding a domain or Google Apps, the correct product prices should be shown everywhere.
- signup: create a site with a domain and Google Apps, check that your cart is correctly populated and can be purchased on checkout
